packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -423,65 +564,128 @@ class FocusScopeNode extends Object with DiagnosticableTreeMixin {
+    }
+    if (_nextFocus == node) {
+      _nextFocus = null;
+      _dirtyNodes.add(node);
+      _markNeedsUpdate();
+    }
   }
 
   bool _haveScheduledUpdate = false;
-  void _markNeedsUpdate() {
-    if (_haveScheduledUpdate)
+  void _markNeedsUpdate({FocusNode newFocus}) {
+    // If newFocus isn't specified, then don't mess with _nextFocus, just
+    // schedule the update.
+    _nextFocus = newFocus ?? _nextFocus;
+    if (_haveScheduledUpdate) {
       return;
+    }
     _haveScheduledUpdate = true;
-    scheduleMicrotask(_update);
+    scheduleMicrotask(_applyFocusChange);
   }
 
-  FocusNode _findNextFocus() {
-    FocusScopeNode scope = rootScope;
-    while (scope._firstChild != null)
-      scope = scope._firstChild;
-    return scope._focus;
-  }
-
-  void _update() {
+  void _applyFocusChange() {
     _haveScheduledUpdate = false;
-    final FocusNode nextFocus = _findNextFocus();
-    if (_currentFocus == nextFocus)
-      return;
     final FocusNode previousFocus = _currentFocus;
-    _currentFocus = nextFocus;
-    previousFocus?._notify();
-    _currentFocus?._notify();
+    if (_currentFocus == null && _nextFocus == null) {
+      _nextFocus = rootScope._policy.findFirstFocus(rootScope);
+    }
+    if (_nextFocus != null && _nextFocus != _currentFocus) {
+      _currentFocus = _nextFocus;
+      final Set<FocusNode> previousPath = previousFocus?.ancestors?.toSet() ?? <FocusNode>{};
+      final Set<FocusNode> nextPath = _nextFocus.ancestors.toSet();
+      // Notify nodes that are newly focused.
+      _dirtyNodes.addAll(nextPath.difference(previousPath));
+      // Notify nodes that are no longer focused
+      _dirtyNodes.addAll(previousPath.difference(nextPath));
+      _nextFocus = null;
+    }
+    if (previousFocus != null) {
+      _dirtyNodes.add(previousFocus);
+    }
+    if (_currentFocus != null) {
+      _dirtyNodes.add(_currentFocus);
+    }
+    for (FocusNode node in _dirtyNodes) {
+      node._notify();
+    }
+    _dirtyNodes.clear();
   }
 
-  List<FocusScopeNode> _getCurrentFocusPath() => _currentFocus?._parent?._getFocusPath();
+  @override
+  List<DiagnosticsNode> debugDescribeChildren() {
+    return <DiagnosticsNode>[
+      rootScope.toDiagnosticsNode(name: 'rootFocusScope'),
+    ];
+  }
 
   @override
-  String toString() {
-    final String status = _haveScheduledUpdate ? ' UPDATE SCHEDULED' : '';
-    const String indent = '  ';
-    return '${describeIdentity(this)}$status\n'
-      '${indent}currentFocus: $_currentFocus\n'
-      '${rootScope.toStringDeep(prefixLineOne: indent, prefixOtherLines: indent)}';
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    properties.add(FlagProperty('haveScheduledUpdate', value: _haveScheduledUpdate, ifTrue: 'UPDATE SCHEDULED'));
+    properties.add(DiagnosticsProperty<FocusNode>('currentFocus', _currentFocus, defaultValue: null));
   }
 }
+
+/// Returns a text representation of the current focus tree.
+String debugFocusTree() {
```

Comment:
should be called debugDumpFocusTree for consistency

Reply:
Done.